meta.jq

```diff
@@ -319,8 +322,50 @@ def build_command:
 
 			# TODO consider / check what "crane validate" does and if it would be appropriate here
 
-			# TODO generate SBOM? ... somehow
-
+			(
+				[
+					# Generate SBOM
+					@sh "SOURCE_DATE_EPOCH=\(.source.entry.SOURCE_DATE_EPOCH)",
+					"docker buildx build --progress=plain",
+					"--provenance=false",
+					"--build-context \"base:latest=oci-layout://$PWD/temp/@$(jq -r '.manifests[0].digest' index.json)\"",
+					"--sbom=generator=\"$BASHBREW_BUILDKIT_SBOM_GENERATOR\"",
+					"--output 'type=oci,tar=false,dest=temp_sbom/'",
+					"- <<<'FROM base'",
+					empty
+				] | join(" \\\n\t")
+			),
+			@sh "newIndexDigest=\"$(jq -r '.manifests[0].digest | " + digest_to_blob + "' temp_sbom/index.json)\"",
+			@sh "sbomManifestDigest=\"$(jq -r '.manifests[] | select(.platform.os == \"unknown\").digest | " + digest_to_blob + "' temp_sbom/blobs/sha256/$newIndexDigest)\"",
+			@sh "sbomManifest=\"$(jq -r --arg digest \"sha256:$sbomManifestDigest\" '.manifests[] | select(.digest == $digest)' temp_sbom/blobs/sha256/$newIndexDigest)\"",
+			@sh "sbomDigest=\"$(jq -r '.layers[0].digest | " + digest_to_blob + "' temp_sbom/blobs/sha256/$sbomManifestDigest)\"",
+			"cp temp_sbom/blobs/sha256/$sbomDigest temp/blobs/sha256",
+			"cp temp_sbom/blobs/sha256/$sbomManifestDigest temp/blobs/sha256",
+			(
+				[
+					"cat <<< $(",
+					(
+						[
+							"\tjq -r",
+							@sh "--argjson manifest \"$sbomManifest\"",
+							(
+								[
+									"'",
+									".manifests[0].digest as $digest |",
+									".manifests[.manifests | length] |= . + $manifest |",
+									".manifests[1].annotations[\"vnd.docker.reference.digest\"] = $digest",
+									"'",
+									empty
+								] | join("\n\t\t")
+							),
+							"temp/index.json",
+							empty
+						] | join(" \\\n\t")
+					),
+					") > temp/index.json",
+					empty
+				] | join(" \\\n")
+			),
```

Comment:
This should _also_ consider `build_should_sbom` so we don't prematurely include SBOMs we didn't intend to :eyes:

Here's a suggested edit that moves things around/adjusts based on my raw script feedback (I took the approach of getting the end result to look the way I wanted it, then worked backwards to making the script create that):

```bash
# ...
mv temp/index.json.new temp/index.json
# SBOM
originalImageManifest="$(jq -r '.manifests[0].digest' temp/index.json)"
SOURCE_DATE_EPOCH=1709081058 \
	docker buildx build --progress=plain \
	--load=false \
	--provenance=false \
	--build-arg BUILDKIT_DOCKERFILE_CHECK=skip=all \
	--sbom=generator="$BASHBREW_BUILDKIT_SBOM_GENERATOR" \
	--build-context "fake=oci-layout://$PWD/temp@$originalImageManifest" \
	--platform 'linux/amd64' \
	--output 'type=oci,tar=false,dest=sbom' \
	- <<<'FROM fake'
sbomIndex="$(jq -r '.manifests[0].digest' sbom/index.json)"
shell="$(jq -r --arg originalImageManifest "$originalImageManifest" '
	first(
		.manifests[]
		| select(.annotations["vnd.docker.reference.type"] == "attestation-manifest")
	) as $attDesc
	| @sh "sbomManifest=\($attDesc.digest)",
		@sh "sbomManifestDesc=\(
			$attDesc
			| .annotations["vnd.docker.reference.digest"] = $originalImageManifest
			| tojson
		)"
' "sbom/blobs/${sbomIndex/://}")"
eval "$shell"
cp "sbom/blobs/${sbomManifest/://}" "temp/blobs/${sbomManifest/://}"
jq -r --argjson sbomManifestDesc "$sbomManifestDesc" '.manifests += [ $sbomManifestDesc ]' temp/index.json > temp/index.json.new
mv temp/index.json.new temp/index.json
# </build>
# ...
```

```suggestion
			if build_should_sbom then
				# we'll trick BuildKit into generating an SBOM for us, then inject it into our OCI layout
				"# SBOM",
				"imageManifest=\"$(jq -r '.manifests[0].digest' temp/index.json)\"",
				(
					[
						@sh "SOURCE_DATE_EPOCH=\(.source.entry.SOURCE_DATE_EPOCH)",
						"docker buildx build --progress=plain",
						"--load=false", "--provenance=false", # explicitly disable a few features we want to avoid
						"--build-arg BUILDKIT_DOCKERFILE_CHECK=skip=all", # disable linting (https://github.com/moby/buildkit/pull/4962)
						"--sbom=generator=\"$BASHBREW_BUILDKIT_SBOM_GENERATOR\"",
						"--build-context \"fake=oci-layout://$PWD/temp@$imageManifest\"",
						@sh "--platform \(.source.arches[.build.arch].platformString)",
						"--output 'type=oci,tar=false,dest=sbom'",
						# TODO also add appropriate "--tag" lines (which would give us a mostly correct "subject" block in the generated SBOM, but we'd then need to replace instances of ${sbomImageManifest#*:} with ${imageManifest#*:} for their values to be correct)
						"- <<<'FROM fake'", # note: "<<<" is a bashism (so this output must be invoked via bash)
						empty
					] | join(" \\\n\t")
				),
				"sbomIndex=\"$(jq -r '.manifests[0].digest' sbom/index.json)\"",
				@sh "shell=\"$(jq -r --arg imageManifest \"$imageManifest\" \("
					# https://docs.docker.com/build/attestations/attestation-storage/
					first(
						.manifests[]
						| select(.annotations[\"vnd.docker.reference.type\"] == \"attestation-manifest\")
					) as $attDesc
					| @sh \"sbomManifest=\\($attDesc.digest)\",
						@sh \"sbomManifestDesc=\\(
							$attDesc
							| .annotations[\"vnd.docker.reference.digest\"] = $imageManifest
							| tojson
						)\"
				" | unindent_and_decomment_jq(4)) \"sbom/blobs/${sbomIndex/://}\")\"",
				"eval \"$shell\"",
				"cp \"sbom/blobs/${sbomManifest/://}\" \"temp/blobs/${sbomManifest/://}\"",
				"jq -r --argjson sbomManifestDesc \"$sbomManifestDesc\" '.manifests += [ $sbomManifestDesc ]' temp/index.json > temp/index.json.new",
				"mv temp/index.json.new temp/index.json",
				empty
			else empty end,

```